rfcs/0003-languages.md

```diff
@@ -0,0 +1,156 @@
+
+# Motivation
+[motivation]: #motivation
+
+As Soxin is aimed at providing its users with easy-to-configure software, while
+allowing for great customization, an RFC describing how such an important
+feature integrates with the project was needed.
+
+# Detailed design
+[design]: #detailed-design
+
+## Language definition
+
+Languages would be defined in a module named `soxin.programming.languages`.
+Each language, would be an attribute set in this module:
+
+```
+soxin.programming.languages = {
+  go = { /* ... */ };
+  java = { /* ... */ };
+  python = { /* ... */ };
+}
+```
+## Tools defination
+
+Programming tools will be defined in a module named `soxin.programming.tools`.
+Each tool will be an attribute set in this module.
+
+```
+soxin.programming.tools = {
+  git = { /* ... */ };
+  tmux = { /* ... */ };
+}
+```
+This allows Soxin to define its own support for languages and tools, while
+allowing room for user customization as well.
+
+Similar to RFC 2, each programming language must include a key for each editor
+which will support that language.  For instance, it could be a plugin for that
+editor that installs support for that language, plus some extra configuration
+to configure that plugin.
+
+## User interaction
+
+### Globally
+
+Similar to RFC 2, a `soxin.settings.programming.languages` and
+`soxin.settings.programming.tools` option would be introduced to allow the user
+to choose his programming stack. The only permitted values here shall be the
+defined in `config.soxin.programming.<type>`. This option shall be of type
+`Array` of `str`, with apply function which iterate over the array and find
+appropriate languages or tools.
+
+### Per-program
+
+Each program that supports a language must provide a
+`soxin.programs.<program-name>.programming.language` and
+`soxin.programs.<editor-name>.programming.tool` option to allow the user to
+override the programming language stack settings per program. These options will default
+to `config.soxin.settings.programming.<type>`, and thus will be of type
+`attrs`. Similar to RFC 2, they will also have will also allow `str` and have
+an `apply` function that will lookup the theme from `config.soxin.programming.<type>`
+for each option.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Here's how one could define a language and a tool implementing only the `neovim`
+editor:
+
+```
+{ pkgs, ... }:
+
+{
+  config.soxin.programming.languages = {
+    go = {
+      neovim = {
+        plugins = [ pkgs.neovim.go ];
+        extraRC = ''
+        '';
+      };
+    };
+
+    config.soxin.programming.tools = {
+      git = {
+        plugins = [ pkgs.neovim.fugitive ];
+        extraRc = ''
+        '';
+      };
+    };
+  };
+}
+```
+
+Here's how this would be implemented on the editor side:
+
+```
+{ config, lib, ... }:
+
+let
+  cfg = config.soxin.programs.neovim;
+  goSupportRc = cfg.programming.languages.go ? "";
```

Comment:
The returned type of this is an attrset, not a string; Shouldn't it be `.go ? {}`?